miniapps/dpg/util/weakform.hpp

```diff
@@ -0,0 +1,246 @@
+      IsTraceFes = false;
+      for (int i = 0; i < nblocks; i++)
+      {
+         IsTraceFes[i] =
+            (dynamic_cast<const H1_Trace_FECollection*>(trial_fes[i]->FEColl()) ||
+             dynamic_cast<const ND_Trace_FECollection*>(trial_fes[i]->FEColl()) ||
+             dynamic_cast<const RT_Trace_FECollection*>(trial_fes[i]->FEColl()));
+      }
+      Init();
+   }
+
+   /// Get the size of the bilinear form of the DPGWeakForm
+   int Size() const { return height; }
+
+   /// Pre-allocate the internal BlockMatrix before assembly.
+   void AllocateMatrix() { if (mat == NULL) { AllocMat(); } }
+
+   ///  Finalizes the matrix initialization.
+   void Finalize(int skip_zeros = 1);
+
+   /// Returns a reference to the BlockMatrix:  \f$ M \f$
+   BlockMatrix &BlockMat()
+   {
+      MFEM_VERIFY(mat, "mat is NULL and can't be dereferenced");
+      return *mat;
+   }
+
+   /// Returns a reference to the sparse matrix of eliminated b.c.: \f$ M_e \f$
+   BlockMatrix &BlockMatElim()
+   {
+      MFEM_VERIFY(mat_e, "mat_e is NULL and can't be dereferenced");
+      return *mat_e;
+   }
+
+   /// Adds new Trial Integrator. Assumes ownership of @a bfi.
+   void AddTrialIntegrator(BilinearFormIntegrator *bfi, int n, int m);
+
+   /// Adds new Test Integrator. Assumes ownership of @a bfi.
+   void AddTestIntegrator(BilinearFormIntegrator *bfi, int n, int m);
+
+   /// Adds new Domain LF Integrator. Assumes ownership of @a bfi.
+   void AddDomainLFIntegrator(LinearFormIntegrator *bfi, int n);
+
+   /// Assembles the form i.e. sums over all integrators.
+   void Assemble(int skip_zeros = 1);
+
+   virtual void FormLinearSystem(const Array<int> &ess_tdof_list, Vector &x,
+                                 OperatorHandle &A, Vector &X,
+                                 Vector &B, int copy_interior = 0);
+
+   template <typename OpType>
+   void FormLinearSystem(const Array<int> &ess_tdof_list, Vector &x,
+                         OpType &A, Vector &X, Vector &B,
+                         int copy_interior = 0)
+   {
+      OperatorHandle Ah;
+      FormLinearSystem(ess_tdof_list, x, Ah, X, B, copy_interior);
+      OpType *A_ptr = Ah.Is<OpType>();
+      MFEM_VERIFY(A_ptr, "invalid OpType used");
+      A.MakeRef(*A_ptr);
+   }
+
+   virtual void FormSystemMatrix(const Array<int> &ess_tdof_list,
+                                 OperatorHandle &A);
+
+   template <typename OpType>
+   void FormSystemMatrix(const Array<int> &ess_tdof_list, OpType &A)
+   {
+      OperatorHandle Ah;
+      FormSystemMatrix(ess_tdof_list, Ah);
+      OpType *A_ptr = Ah.Is<OpType>();
+      MFEM_VERIFY(A_ptr, "invalid OpType used");
+      A.MakeRef(*A_ptr);
+   }
+
+   void EliminateVDofs(const Array<int> &vdofs,
+                       Operator::DiagonalPolicy dpolicy = Operator::DIAG_ONE);
+
+   void EliminateVDofsInRHS(const Array<int> &vdofs, const Vector &x, Vector &b);
+
+   virtual void RecoverFEMSolution(const Vector &X,Vector &x);
```

Comment:
Please briefly document the public functions here that users will call. If there are similar functions in existing analogous classes like `BilinearForm`, it would be nice to copy the similar documentation there as well.

Reply:
done!